### PR TITLE
Kevin async question verify notifications

### PR DESCRIPTION
**Pull request by wskksw**
# Description

1. allow students to determine whether they want to escalate questions
- After posting a question, AI attempts to answer the question, student can then choose whether they are satisfied with the answer.
- These questions flashes for the creator to suggest that they need action from students.

2. Render Async questions of different status with different UI (human verified vs pending)
- Only human answered ones are green background. 

3. Differentiate Async question status further
- question status enum changed
- async question now has visible (to all students) and verified column

4. Added integration testing for routes for async questions and passes
5. Document logic of new async question feature

Bug fixes/improvements:
1. add taHelpedId in question answering
2. change question filtering to reflect deleted questions
3. trigger asyncquestions to refetch questions upon edit/creation of questions

Closes # (issue number)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [ ] This requires a run of `yarn install`

# How Has This Been Tested?

Please describe how you tested this PR (both manually and with tests)
Provide instructions so we can reproduce. 

- [x] Integration test of backend

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code where needed 
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/app/components/Questions/AsyncQuestions/AsyncQuestionForm.tsx
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/AsyncQuestionForm.tsx:1]: When creating a question the form is populated with previous question. Is this intentional?
Steps to reproduce:

1. **As Student** => Create a new question
2. **As Student** => Click "Post your Question"
3. The form is pre-populated with content from first question

File: packages/app/components/Questions/AsyncQuestions/AsyncQuestionForm.tsx [hunk 1]
```diff
@@ -93,21 +82,21 @@ export function AsyncQuestionForm({
...
-    let questionId = NaN
+    const aiAnswer = await getAiAnswer(
+      value.QuestionAbstract + ' ' + value.questionText,
+    )
     await API.asyncQuestions
       .create(
         {
           questionTypes: questionTypeInput,
           questionText: value.questionText,
-          aiAnswerText: 'AI answer being generated...',
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/AsyncQuestionForm.tsx:102]: Why did you get rid of this?

File: packages/app/components/Questions/AsyncQuestions/AsyncQuestions.tsx [hunk 2]
```diff
@@ -76,11 +77,17 @@ export default function AsyncQuestionsPage({
   const { questions, mutateQuestions } = useAsnycQuestions(courseId)
 
   useEffect(() => {
+    console.log(questions)
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/AsyncQuestions.tsx:80]: remove this?

File: packages/app/components/Questions/AsyncQuestions/AsyncQuestions.tsx [hunk 3]
```diff
@@ -120,6 +129,7 @@ export default function AsyncQuestionsPage({
       .map((question) => question.questionTypes)
       .flat()
     setQuestionsTypeState(shownQuestionTypes)
+    console.log(displayedQuestions)
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/AsyncQuestions.tsx:132]: remove this?

File: packages/app/components/Questions/AsyncQuestions/StudentQuestionDetailButtons.tsx [hunk 4]
```diff
@@ -11,38 +11,45 @@ import { AsyncQuestionForm } from './AsyncQuestionForm'
...
   const [answerQuestionVisible, setAnswerQuestionVisbile] = useState(false)
   // const handleCancel = () => {
   //   setAnswerQuestionVisbile(false);
   // };
   // const [form] = Form.useForm();
 
-  if (question.status !== asyncQuestionStatus.Waiting) {
-    return <></>
-  }
+  // if (question.status !== asyncQuestionStatus.Waiting) {
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/StudentQuestionDetailButtons.tsx:26]: Remove if not needed

File: packages/app/components/Questions/AsyncQuestions/TAanswerQuestionModal.tsx [hunk 5]
```diff
@@ -21,13 +30,15 @@ export function AnswerQuestionModal({
   //use questions for form validation
   useEffect(() => {
     form.setFieldsValue(question)
+    console.log(question)
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/TAanswerQuestionModal.tsx:33]: remove this

File: packages/app/components/Questions/AsyncQuestions/TAanswerQuestionModal.tsx [hunk 6]
```diff
@@ -87,6 +98,24 @@ export function AnswerQuestionModal({
           visible: question.visible,
         }}
       >
+        <Form.Item
+          name="visible"
+          label="Set question visible to all students"
+          valuePropName="checked"
+        >
+          <Switch
```
Comment (Ferdinand737) [inline packages/app/components/Questions/AsyncQuestions/TAanswerQuestionModal.tsx:106]: I really like these Switches

File: packages/server/src/asyncQuestion/asyncQuestion.entity.ts
Comment (Ferdinand737) [inline packages/server/src/asyncQuestion/asyncQuestion.entity.ts:1]: I do not see a migration file for this change. 
`yarn migration:generate -n add-verified-to-async`

File: packages/server/src/asyncQuestion/asyncQuestion.controller.ts [hunk 7]
```diff
@@ -66,22 +67,24 @@ export class asyncQuestionController {
...
       //   },
       // });
-      // console.log(professors);
-      // const post: sendEmailAsync = {
-      //   receiver: professors.user.email,
-      //   subject: 'UBC helpme Async question created',
-      //   type: asyncQuestionEventType.created,
-      // };
-      // this.mailService.sendEmail(post);
+      // faculties.forEach((professors) => {
```
Comment (Ferdinand737) [inline packages/server/src/asyncQuestion/asyncQuestion.controller.ts:80]: Remove this if not needed

File: packages/server/src/asyncQuestion/asyncQuestion.controller.ts [hunk 8]
```diff
@@ -111,29 +115,52 @@ export class asyncQuestionController {
...
+      // };
+      // if (body.status === asyncQuestionStatus.HumanAnswered) {
+      //   post.type = asyncQuestionEventType.answered;
+      //   this.mailService.sendEmail(post);
+      // } else if (body.status === asyncQuestionStatus.TADeleted) {
+      //   post.type = asyncQuestionEventType.deleted;
+      //   this.mailService.sendEmail(post);
+      // }
     });
     return question;
```
Comment (Ferdinand737) [inline packages/server/src/asyncQuestion/asyncQuestion.controller.ts:165]: consider using `res.send()`

**Review by Ferdinand737 — CHANGES_REQUESTED** [review]
I found a bug while testing that creates duplicates.

image


Steps to reproduce.

1. **As Student** => Create question
2. **As Student** => Do not click 'satisfied' or 'need faculty help'
3. **As Prof** => Post response to question and mark "verified" and "visible"

**Comment by wskksw** [comment]
The bug with duplicates is found, it turns out that the same question is returned twice both in helpedQuestions and visibleQuestions filtering, I would do a bit more to see what's the best way to filter and fix the bug.

**Comment by wskksw** [comment]
@Ferdinand737 Have completed requested changes, lmk whether they suffice

**Comment by AdamFipke** [comment]
While I like the yellow background for when the question still needs attention by the student (i.e. if they're satisfied or not), I **think it should be a white background if the student is satisfied.** I think the background should only be yellow if either:
- the student hasn't yet said responded if they're satisfied/not satisfied
- the question needs attention from the staff (i.e. student not satisfied)

I like the yellow background since it will help bring attention to questions that probably need to be edited (if a professor) or are likely to get updated in the future (if a student) 

image

File: packages/app/components/Questions/AsyncQuestions/README.MD [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+
+## Expanded async cards
+
+By default, all questions are collapsed. However, there are a few scenarios where a question card will be expanded:
+- clicking on a card will expand the card to show/hide the answer
+- after creating a question, the card will be expanded to show the answer
+- after editing a question, the card will be expanded to show the answer
+
+Due to the variability of expanded or not, a state in asyncquestions.tsx is maintained to keep track of which cards are expanded or not by storing the question id in an array/variable.
```
Comment (AdamFipke) [inline packages/app/components/Questions/AsyncQuestions/README.MD:33]: > Due to the variability of expanded or not, a state in asyncquestions.tsx is maintained to keep track of which cards are expanded or not by storing the question id in an array/variable.

This is helpful, I like this kind of documentation that explains why something is done a certain way

**Review by AdamFipke — APPROVED** [review]
I did a little test of it manually, I think it works good.

Though, seeing as how @Ferdinand737 also did a bunch of work on the async questions, I think we should wait to merge this until he approves.

**Comment by Ferdinand737** [comment]
Im getting a really strange issue. 
Steps to reproduce:
1. Log in as student and create async question
2. Log in as instructor and go to `http://localhost:3000/course/1/async_question`
3. The page loads forever and does not display the questions browser becomes unresponsive

I tried this with multiple browsers, firefox gave this message.
image

**Comment by wskksw** [comment]
I can't reproduce the strange issue you mentioned. Can someone else try it too?

**Comment by Ferdinand737** [comment]
> I can't reproduce the strange issue you mentioned. Can someone else try it too?

It works now. I dont know what happened

**Review by Ferdinand737 — APPROVED** [review]

**Pull request merged**